File: app/src/main/java/com/karol/sezonnazdrowie/view/fragments/SettingsItemsFragment.java
package com.karol.sezonnazdrowie.view.fragments;

import android.content.SharedPreferences;
import android.os.Bundle;

import com.karol.sezonnazdrowie.R;
import com.karol.sezonnazdrowie.data.FoodItem;
import com.karol.sezonnazdrowie.model.MainViewModel;
import com.karol.sezonnazdrowie.view.MainActivity;

import java.util.List;

import androidx.lifecycle.ViewModelProviders;
import androidx.preference.CheckBoxPreference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceScreen;

public class SettingsItemsFragment extends PreferenceFragmentCompat
        implements SharedPreferences.OnSharedPreferenceChangeListener {

    private MainViewModel mMainViewModel;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        mMainViewModel = ViewModelProviders.of(getActivity()).get(MainViewModel.class);
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.prefs_list, rootKey);

        PreferenceScreen screen = this.getPreferenceScreen();
        List<FoodItem> list = null;
        if (getArguments().getString(MainActivity.INTENT_WHAT)
                .equals(MainActivity.INTENT_WHAT_FRUITS)) {
            list = mMainViewModel.getDatabase().getAllFruits();
        } else if (getArguments().getString(MainActivity.INTENT_WHAT)
                .equals(MainActivity.INTENT_WHAT_VEGETABLES)) {
            list = mMainViewModel.getDatabase().getAllVegetables();
        }

        for (FoodItem item : list) {
            if (item.getStartDay1() == null) {
                continue;
            }
            CheckBoxPreference pref = new CheckBoxPreference(screen.getContext());
            pref.setTitle(item.getName());
            pref.setKey("pref_noti_" + item.getName());
            pref.setDefaultValue(true);
            screen.addPreference(pref);
        }

        mMainViewModel.setSettingsItemChanged(false);
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        super.onPause();
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {
        mMainViewModel.setSettingsItemChanged(true);
    }
}

File: app/src/main/java/com/karol/sezonnazdrowie/view/fragments/SettingsItemsFragment.kt
package com.karol.sezonnazdrowie.view.fragments

import android.content.SharedPreferences
import android.os.Bundle
import androidx.lifecycle.ViewModelProviders
import androidx.preference.CheckBoxPreference
import androidx.preference.PreferenceFragmentCompat
import com.karol.sezonnazdrowie.R
import com.karol.sezonnazdrowie.model.MainViewModel
import com.karol.sezonnazdrowie.view.MainActivity
import com.karol.sezonnazdrowie.view.MainActivity.Companion.INTENT_WHAT
import java.util.concurrent.atomic.AtomicBoolean

class SettingsItemsFragment : PreferenceFragmentCompat(),
    SharedPreferences.OnSharedPreferenceChangeListener {

    private lateinit var mainViewModel: MainViewModel

    private val listenerRegistered = AtomicBoolean(false)

    override fun onCreate(savedInstanceState: Bundle?) {
        mainViewModel = ViewModelProviders.of(activity!!).get(MainViewModel::class.java)
        super.onCreate(savedInstanceState)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.prefs_list, rootKey)

        mainViewModel.isSettingsItemChanged = false
        if (listenerRegistered.compareAndSet(false, true)) {
            preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
        }

        val screen = this.preferenceScreen
        val list = when (val intentWhat = arguments?.getString(INTENT_WHAT)
            ?: throw IllegalArgumentException("Missing arguments!")) {
            MainActivity.INTENT_WHAT_FRUITS ->
                mainViewModel.database.allFruits.filter { !it.isFullYear() }
            MainActivity.INTENT_WHAT_VEGETABLES ->
                mainViewModel.database.allVegetables.filter { !it.isFullYear() }
            else -> throw IllegalArgumentException("Unknown type of page $intentWhat")
        }

        list.filter { it.startDay1 != null }
            .map { it.name }
            .forEach { name ->
                CheckBoxPreference(screen.context)
                    .apply {
                        title = name
                        key = "pref_noti_$name"
                        isIconSpaceReserved = false
                        setDefaultValue(true)
                    }.also {
                        screen.addPreference(it)
                    }
            }
    }

    override fun onResume() {
        super.onResume()
        if (listenerRegistered.compareAndSet(false, true)) {
            preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
        }
    }

    override fun onPause() {
        if (listenerRegistered.compareAndSet(true, false)) {
            preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        }
        super.onPause()
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, s: String) {
        mainViewModel.isSettingsItemChanged = true
    }
}
